Math/NumberTheory/SmoothNumbers.hs

```diff
@@ -76,29 +79,36 @@ fromSmoothUpperBound n = if (n < 2)
                          then Nothing
                          else Just $ SmoothBasis $ takeWhile (<= n) $ map unPrime primes
 
--- | Generate an infinite ascending list of
--- <https://en.wikipedia.org/wiki/Smooth_number smooth numbers>
--- over a given smooth basis.
---
--- >>> import Data.Maybe
--- >>> take 10 (smoothOver (fromJust (fromList [2, 5])))
--- [1, 2, 4, 5, 8, 10, 16, 20, 25, 32]
-smoothOver :: Integral a => SmoothBasis a -> [a]
-smoothOver pl = foldr (\p l -> mergeListLists $ iterate (map (p*)) l) [1] (unSmoothBasis pl)
+-- | Helper used by @smoothOver@. Since the typeclass constraint is just
+-- @Euclidean@, with Euclidean domains not being required to have a total
+-- ordering defined, it receives a @norm@ comparison function for the
+-- generated smooth numbers.
+smoothOver' :: forall a b . (E.Euclidean a, Ord b) => (a -> b) -> SmoothBasis a -> [a]
+smoothOver' norm pl = foldr (\p l -> mergeListLists $ iterate (map (p*)) l) [1] (unSmoothBasis pl)
   where
     {-# INLINE mergeListLists #-}
     mergeListLists      = foldr go1 []
       where
-        go1 :: Ord a => [a] -> [a] -> [a]
+        go1 :: [a] -> [a] -> [a]
         go1 (h:t) b = h:(go2 t b)
         go1 _     b = b
 
-        go2 :: Ord a => [a] -> [a] -> [a]
+        go2 :: [a] -> [a] -> [a]
         go2 a@(ah:at) b@(bh:bt)
-          | bh < ah   = bh : (go2 a bt)
+          | norm bh < norm ah   = bh : (go2 a bt)
           | otherwise = ah : (go2 at b) -- no possibility of duplicates
         go2 a b = if null a then b else a
 
+-- | Generate an infinite ascending list of
+-- <https://en.wikipedia.org/wiki/Smooth_number smooth numbers>
+-- over a given smooth basis.
+--
+-- >>> import Data.Maybe
+-- >>> take 10 (smoothOver (fromJust (fromList [2, 5])))
+-- [1, 2, 4, 5, 8, 10, 16, 20, 25, 32]
+smoothOver :: (E.Euclidean a, Ord a) => SmoothBasis a -> [a]
+smoothOver = smoothOver' abs
```

Comment:
The problem with `abs` is that it does not satisfy a prerequisite of `smoothOver'` stated in the comment above. For instance, `abs (2 + ι) > abs ((2 + ι) * (1 + ι))`. It means that `mergeListLists` is fed with unsorted (in terms of `abs`) lists and consequently returns unsorted (again in terms of `abs`) results.

For example, notice an unexpected position of `3+4*ι`:
```haskell
> take 10 $ map abs $ smoothOver $ fromJust $ fromList [1+ι, 2+ι]
[1,1+ι,1+3*ι,2,2+ι,2+2*ι,2+6*ι,3+4*ι,2+11*ι,4]
```

I suggest to leave `smoothOver` constrained by `Integral a`, in which case `abs` is a valid norm. And expose `smoothOver'` to users as a function, which works for any `Euclidean`.